gossip/emitter/throttler/throttler.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2025 Sonic Operations Ltd
+// This file is part of the Sonic Client
+//
+// Sonic is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Sonic is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with Sonic. If not, see <http://www.gnu.org/licenses/>.
+
+package throttler
+
+import (
+	"github.com/0xsoniclabs/sonic/gossip/emitter/config"
+	"github.com/0xsoniclabs/sonic/inter"
+	"github.com/0xsoniclabs/sonic/opera"
+	"github.com/Fantom-foundation/lachesis-base/inter/idx"
+	"github.com/Fantom-foundation/lachesis-base/inter/pos"
+)
+
+//go:generate mockgen -source=throttler.go -destination=throttler_mock.go -package=throttler
+
+type WorldReader interface {
+	GetRules() opera.Rules
+	GetEpochValidators() (*pos.Validators, idx.Epoch)
+	GetLastEvent(idx.ValidatorID) *inter.Event
+}
+
+// validatorAttendance holds information about a validator's online status.
+type validatorAttendance struct {
+	lastSeenSeq idx.Event
+	lastSeenAt  config.Attempt
+
+	online bool
+}
+
+// dominantSet is a helper tool to compute the set of online validators.
+type attendanceList map[idx.ValidatorID]validatorAttendance
+
+// updateAttendance updates the attendance list based on the current world state and configuration.
+func (al attendanceList) updateAttendance(
+	world WorldReader, config config.ThrottlerConfig,
+	lastDominantSet dominantSet, attempt config.Attempt) {
+
+	validators, _ := world.GetEpochValidators()
+	for _, id := range validators.IDs() {
+
+		lastEvent := world.GetLastEvent(id)
+		if lastEvent == nil {
+			// No event has been seen from this validator yet
+			continue
+		}
+
+		attendance, exists := al[id]
+
+		// Different tolerance for being online for dominant vs non-dominant validators.
+		// Relaxed tolerance can only be used if the validator was online previously.
+		onlineThreshold := config.DominatingTimeout
+		if exists && attendance.online {
+			if _, wasDominant := lastDominantSet[id]; !wasDominant {
+				onlineThreshold = config.NonDominatingTimeout
+			}
+		}
```

Comment:
This logic seems like something that should be handled by the config, maybe make it a function there that takes a validator id and returns the corresponding timeout for it?